primitives/src/merkle_tree/namespaced_merkle_tree.rs

```diff
@@ -0,0 +1,262 @@
+where
+    N: Namespace,
+    T: NodeValue,
+{
+    min_namespace: N,
+    max_namespace: N,
+    hash: T,
+}
+
+impl<T, N> Default for NamespacedHash<T, N>
+where
+    N: Namespace,
+    T: NodeValue,
+{
+    fn default() -> Self {
+        Self {
+            hash: T::default(),
+            max_namespace: <N as Namespace>::min(),
+            min_namespace: <N as Namespace>::max(),
+        }
+    }
+}
+
+impl<T, N> NamespacedHash<T, N>
+where
+    N: Namespace,
+    T: NodeValue,
+{
+    /// Constructs a new NamespacedHash
+    pub fn new(min_namespace: N, max_namespace: N, hash: T) -> Self {
+        Self {
+            min_namespace,
+            max_namespace,
+            hash,
+        }
+    }
+}
+
+impl<E, H, T, I, N> DigestAlgorithm<E, I, NamespacedHash<T, N>> for NamespacedHasher<H, E, I, T, N>
+where
+    E: Element + Namespaced<N>,
+    I: Index,
+    N: Namespace,
+    T: NodeValue,
+    H: DigestAlgorithm<E, I, T> + BindNamespace<E, I, T, N>,
+{
+    fn digest(data: &[NamespacedHash<T, N>]) -> NamespacedHash<T, N> {
+        if data.is_empty() {
+            return NamespacedHash::default();
+        }
+        let first_node = data[0];
+        let min_namespace = first_node.min_namespace;
+        let mut max_namespace = first_node.max_namespace;
+        let mut nodes = vec![H::generate_namespaced_commitment(first_node)];
+        for node in data {
+            // Ensure that namespaced nodes are sorted
+            if node.min_namespace < max_namespace {
+                panic!("leaves are out of order")
+            }
+            max_namespace = node.max_namespace;
+            nodes.push(H::generate_namespaced_commitment(*node));
+        }
+
+        let inner_hash = H::digest(&nodes);
+
+        NamespacedHash::new(min_namespace, max_namespace, inner_hash)
+    }
+
+    fn digest_leaf(pos: &I, elem: &E) -> NamespacedHash<T, N> {
+        let namespace = elem.get_namespace();
+        let hash = H::digest_leaf(pos, elem);
+        NamespacedHash::new(namespace, namespace, hash)
+    }
+}
+
+#[cfg(test)]
+mod nmt_tests {
+    use digest::Digest;
+    use sha3::Sha3_256;
+
+    use super::*;
+    use crate::merkle_tree::examples::{Sha3Digest, Sha3Node};
+
+    type NamespaceId = u64;
+    type Hasher = NamespacedHasher<Sha3Digest, Leaf, NamespaceId, Sha3Node, u64>;
+
+    #[derive(
+        Default,
+        Eq,
+        PartialEq,
+        Hash,
+        Ord,
+        PartialOrd,
+        Copy,
+        Clone,
+        Debug,
+        CanonicalSerialize,
+        CanonicalDeserialize,
+    )]
+    struct Leaf {
+        namespace: NamespaceId,
+    }
+
+    impl Namespaced<NamespaceId> for Leaf {
+        fn get_namespace(&self) -> NamespaceId {
+            self.namespace
+        }
+    }
+
+    impl<E, I, N> BindNamespace<E, I, Sha3Node, N> for Sha3Digest
+    where
+        E: Element + CanonicalSerialize,
+        I: Index,
+        N: Namespace,
+    {
+        fn generate_namespaced_commitment(
```

Comment:
As discussed, in practice we should be careful about how we hash min,max and content together. Maybe keeping min and max value without hashing them is fine.

Reply:
If they aren’t hashed then I believe a dishonest prover can modify namespace ranges and return incomplete namespaces

Reply:
What I meant is that we keep bot min and max value as part of the node, which means they are conceptually hashed, the hash being `(min,max,Hash(...) )` . The output of this hash is larger but it is still a (collision-resistant) hash.